File: app/src/main/java/com/termux/view/textselection/TextSelectionHandleView.java
package com.termux.view.textselection;

import android.graphics.Canvas;
import android.view.MotionEvent;
import android.view.View;
import android.widget.PopupWindow;

import com.termux.NyxActivity;
import com.termux.utils.Theme;
import com.termux.view.Console;


final class TextSelectionHandleView extends View {
    private final PopupWindow mHandle = new PopupWindow(this, 40, 40);
    private final int cur;
    private final int n;
    private final TextSelectionCursorController ts;
    private float dx;
    private float dy;


    TextSelectionHandleView(final int num, final TextSelectionCursorController tsc) {
        super(NyxActivity.console.getContext());
        this.n = num;
        this.cur = 0 == num ? 2 : 0;
        this.ts = tsc;
    }

    public void hide() {
        this.mHandle.dismiss();
        TextSelectionCursorController.selectors[this.n] = TextSelectionCursorController.selectors[this.n + 1] = -1;
    }

    public void positionAtCursor(final int cx, final int cy) {
        TextSelectionCursorController.selectors[this.n] = cx;
        TextSelectionCursorController.selectors[this.n + 1] = cy;
        this.update();
    }

    public void update() {
        final var x = NyxActivity.console.getPointX(TextSelectionCursorController.selectors[this.n]) + TextSelectionCursorController.consoleCord[0];
        final var y = NyxActivity.console.getPointY(TextSelectionCursorController.selectors[this.n + 1] + 1) + TextSelectionCursorController.consoleCord[1];
        if (this.mHandle.isShowing()) this.mHandle.update(x, y, -1, -1);
        else this.mHandle.showAtLocation(NyxActivity.console, 0, x, y);
    }

    @Override
    protected void onDraw(final Canvas c) {
        c.drawColor(Theme.primary);
    }

    @Override
    public boolean onTouchEvent(final MotionEvent event) {
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN -> {
                this.dx = event.getX();
                this.dy = event.getY();
                this.ts.hideFloatingMenu();
            }

            case MotionEvent.ACTION_MOVE ->
                    this.updatePosition(event.getRawX() - this.dx - TextSelectionCursorController.consoleCord[0], event.getRawY() - this.dy - TextSelectionCursorController.consoleCord[1]);

            case MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> this.ts.showFloatingMenu();
        }
        return true;
    }


    private void updatePosition(final float x, final float y) {
        final var screen = NyxActivity.console.mEmulator.screen;
        final var scrollRows = screen.activeRows() - NyxActivity.console.mEmulator.mRows;
        TextSelectionCursorController.selectors[this.n] = Console.getCursorX(x);
        TextSelectionCursorController.selectors[this.n + 1] = NyxActivity.console.getCursorY(y);
        if (0 > TextSelectionCursorController.selectors[this.n])
            TextSelectionCursorController.selectors[this.n] = 0;

        if (TextSelectionCursorController.selectors[this.n + 1] < -scrollRows)
            TextSelectionCursorController.selectors[this.n + 1] = -scrollRows;
        else if (TextSelectionCursorController.selectors[this.n + 1] > NyxActivity.console.mEmulator.mRows - 1)
            TextSelectionCursorController.selectors[this.n + 1] = NyxActivity.console.mEmulator.mRows - 1;

        if (TextSelectionCursorController.selectors[1] > TextSelectionCursorController.selectors[3])
            TextSelectionCursorController.selectors[this.n + 1] = TextSelectionCursorController.selectors[this.cur + 1];

        if (TextSelectionCursorController.selectors[1] == TextSelectionCursorController.selectors[3] && TextSelectionCursorController.selectors[0] > TextSelectionCursorController.selectors[2])
            TextSelectionCursorController.selectors[this.n] = TextSelectionCursorController.selectors[this.cur];

        if (!NyxActivity.console.mEmulator.isAlternateBufferActive()) {
            var topRow = NyxActivity.console.topRow;
            if (TextSelectionCursorController.selectors[this.n + 1] <= topRow) {
                topRow--;
                if (topRow < -scrollRows) topRow = -scrollRows;

            } else if (TextSelectionCursorController.selectors[this.n + 1] >= topRow + NyxActivity.console.mEmulator.mRows) {
                topRow++;
                if (0 < topRow) topRow = 0;
            }
            NyxActivity.console.topRow = topRow;
        }
        NyxActivity.console.invalidate();
        this.update();
    }
}

File: app/src/main/java/com/termux/view/textselection/TextSelectionHandleView.kt
package com.termux.view.textselection

import android.graphics.Canvas
import android.graphics.Paint
import android.view.MotionEvent
import android.view.View
import android.widget.PopupWindow
import com.termux.utils.data.TerminalManager.console
import com.termux.utils.ui.primary
import com.termux.view.textselection.TextSelectionCursorController.consoleCord
import com.termux.view.textselection.TextSelectionCursorController.hideFloatingMenu
import com.termux.view.textselection.TextSelectionCursorController.showFloatingMenu

private val paint by lazy { Paint().apply { color = primary } }

class TextSelectionHandleView(val int: Int) : View(console.context) {
    private val mHandle = PopupWindow(this, 40, 40)
    private val cur = if (int == 0) 2 else 0

    fun hide() {
        mHandle.dismiss()
        selectors[int] = -1
        selectors[int + 1] = -1
    }


    fun positionAtCursor(cx: Int, cy: Int) {
        selectors[int] = cx
        selectors[int + 1] = cy
        update()
    }

    fun update() {
        val x = console.getPointX(selectors[int]) + consoleCord[0]
        val y = console.getPointY(selectors[int + 1] + 1) + consoleCord[1]
        if (isShowing) mHandle.update(x, y, -1, -1)
        else mHandle.showAtLocation(console, 0, x, y)
    }


    public override fun onDraw(c: Canvas): Unit = c.drawCircle(20f, 20f, 20f, paint)

    private var dx = 0f
    private var dy = 0f
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                dx = event.x
                dy = event.y
                hideFloatingMenu()
            }

            MotionEvent.ACTION_MOVE -> {
                updatePosition(event.rawX - dx - consoleCord[0], event.rawY - dy - consoleCord[1])
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> showFloatingMenu()
        }
        return true
    }

    private val isShowing: Boolean
        get() = mHandle.isShowing

    private fun updatePosition(x: Float, y: Float) {
        val screen = console.mEmulator.screen
        val scrollRows = screen.activeRows - console.mEmulator.mRows
        selectors[int] = console.getCursorX(x)
        selectors[int + 1] = console.getCursorY(y)
        if (0 > selectors[int]) {
            selectors[int] = 0
        }
        if (selectors[int + 1] < -scrollRows) {
            selectors[int + 1] = -scrollRows
        } else if (selectors[int + 1] > console.mEmulator.mRows - 1) {
            selectors[int + 1] = console.mEmulator.mRows - 1
        }
        if (selectors[1] > selectors[3]) {
            selectors[int + 1] = selectors[cur + 1]
        }
        if (selectors[1] == selectors[3] && selectors[0] > selectors[2]) {
            selectors[int] = selectors[cur]
        }
        if (!console.mEmulator.isAlternateBufferActive) {
            var topRow = console.topRow
            if (selectors[int + 1] <= topRow) {
                topRow--
                if (topRow < -scrollRows) {
                    topRow = -scrollRows
                }
            } else if (selectors[int + 1] >= topRow + console.mEmulator.mRows) {
                topRow++
                if (0 < topRow) {
                    topRow = 0
                }
            }
            console.topRow = topRow
        }
        console.invalidate()
        update()
    }
}
